python/ccf/clients.py

```diff
@@ -81,6 +82,60 @@ def makefile(self, *args, **kwargs):
         return self.file
 
 
+class ResponseBody(abc.ABC):
+    @abc.abstractmethod
+    def data(self) -> bytes:
+        pass
+
+    @abc.abstractmethod
+    def text(self) -> str:
+        pass
+
+    @abc.abstractmethod
+    def json(self) -> Any:
+        pass
+
+    def __len__(self):
+        return len(self.data())
+
+    def __str__(self):
+        try:
+            return self.text()
+        except UnicodeDecodeError:
```

Comment:
I don't think we should catch that, if a user asks for text, they should get the UnicodeDecodeError and handle it how they want, this is making it very difficult for them to do so.

Reply:
I don't think `str(..)` is ever supposed to throw?

Reply:
No? I can't find a reference either way, this isn't mentioned in https://docs.python.org/3/reference/datamodel.html#object.__str__

At least in 2.x, it definitely used to happen on encoding failures, but that particular behaviour is gone from 3.x.

Could we default to __repr__ on exception perhaps?

Reply:
I added the repr fallback.